dali/test/python/test_dali_variable_batch_size.py

```diff
@@ -0,0 +1,622 @@
+import numpy as np
+import test_utils
+import inspect
+import os
+import math
+
+"""
+How to test variable (iter-to-iter) batch size for a given op?
+-------------------------------------------------------------------------------
+The idea is to create a Pipeline that assumes i2i variability, run 2 iterations
+and compare them with ad-hoc created Pipelines for given (constant) batch sizes.
+This can be easily done using `check_batch` function below.
+
+On top of that, there are some utility functions and routines to help with some
+common cases:
+1. If the operator is typically processing image-like data (i.e. 3-dim, uint8,
+   0-255, with shape like [640, 480, 3]) and you want to test default arguments
+   only, please add a record to the `ops_image_default_args` list
+2. If the operator is typically processing image-like data (i.e. 3-dim, uint8,
+   0-255, with shape like [640, 480, 3]) and you want to specify any number of
+   its arguments, please add a record to the `ops_image_custom_args` list
+3. If the operator is typically processing audio-like data (i.e. 1-dim, float,
+   0.-1.) please add a record to the `float_array_ops` list
+4. If your operator case doesn't fit any of the above, please create a nosetest
+   function, in which you can define a function, that returns not yet built
+   pipeline, and pass it to the `check_batch` function.
+5. If your operator performs random operation, this approach won't provide
+   a comparable result. In this case, the best thing you can do is to check
+   whether the operator works, without qualitative comparison. Use `run_pipeline`
+   instead of `check_pipeline`. 
+"""
+
+
+def generate_data(max_batch_size, n_iter, sample_shape, lo=0., hi=1., dtype=np.float32):
+    """
+    Generates an epoch of data, that will be used for variable batch size verification.
+
+    :param max_batch_size: Actual sizes of every batch in the epoch will be less or equal to max_batch_size
+    :param n_iter: Number of iterations in the epoch
+    :param sample_shape: If sample_shape is callable, shape of every sample will be determined by
+                         calling sample_shape. In this case, every call to sample_shape has to
+                         return a tuple of integers. If sample_shape is a tuple, this will be a
+                         shape of every sample.
+    :param lo:
+    :param hi:
+    :param dtype: 'int' for uint8 or 'float' for float32
+    :return: An epoch of data
+    """
+    batch_sizes = np.array([max_batch_size // 2, max_batch_size // 4, max_batch_size])
+
+    if isinstance(sample_shape, tuple):
+        size = sample_shape
+    elif inspect.isgeneratorfunction(sample_shape):
+        size = sample_shape()
+    else:
+        raise RuntimeError(
+            "`sample_shape` shall be either a tuple or a callable. Provide `(val,)` tuple for 1D shape")
+
+    if np.issubdtype(dtype, np.integer):
+        return [np.random.randint(lo, hi, size=(1, bs) + size, dtype=dtype) for bs in
```

Comment:
```suggestion
        return [np.random.randint(lo, hi, size=(bs) + size, dtype=dtype) for bs in
```
You don't need the size=1 trailing dimension.

Reply:
Done